### ISSUE TITLE
Long term compatibility plans

### ISSUE THREAD

**Issue by uramer** [NepEgor/OpenTrackpadController#1]
Hey. Your project looks great, and I'm curious how you intend to make it supported by various games and applications.

I see currently you are using XInput, which is obviously not ideal, as that doesn't support either gyro or touchpads. Have you considered how else to enable all of the controller's features?

One option is simulating Playstation controllers, although that's still awkward due to touchpad position being very different.  
I wonder if it would be possible to create an SDL controller profile, or support Steam Input somehow.

**Comment by NepEgor** [comment]
Hello. Thank you for your interest in the project.

I am planning to implement onboard configuration storage. The controller would have multiple profiles that it could switch based on the game. I am using xinput now because that's how my first target game (Genshin Imapct) works best for me. I will be adding mouse & keyboard emulation support in the future.

Right now gyro and trackpads are converted to xinput joysticks on board of the controller. And the game sees the controller as an xbox controller. I have released a showcase video on YouTube yesterday and in the video you can see how gyro is mixed with right trackpad into xinput right joystick. Please check it out [here](https://www.youtube.com/watch?v=mnsVAtMxI3E&ab_channel=EgorNep).

I am not sure what an SDL controller profile is. Does it require a game developer to use SDL? If so than it would limit the games my controller is usable with.
The same goes for SIAPI. Not many games implement it natively and the games that doesn't are supported with legacy SIAPI. Which is xinput + m&kb - what I am planning to use.

**Comment by uramer** [comment]
Do you intend to support mixed input (e. g. XInput + keyboard + mouse, like Steam Input)?

Yeah, I've seen the video.

An SDL controller profile maps specific device's inputs to a set of standard controller inputs. I. e. it allows the games using SDL to reference Xbox controller dpads, triggers, sticks etc the same as they would Playstation controller ones. Quite a few games use SDL, and it's also recommended as the input library by Valve. I'm suggesting this for native controller support, and realistically only games using SDL could opt for that. Having the controller work with SDL would also mean creating custom input wrappers on top of it is easier. Naturally you'd need other options to make the controller work in all games.

Steam Input API is indeed not very popular, but if there was a way to plug into Steam Input somehow, it would allow you to reuse all the input customization tools from it, which are also familiar to users. It might even be feasible to reuse game controller configurations for Steam Controller.

**Comment by NepEgor** [comment]
Yes I intend to support mixed input on the controller level. Still not every game supports it unfortunately...

I want to first implement the solution that works well for all games. And I think xinput is supported by most of them. SDL seems like an edge case to me that requires too much for what it will give...

SIAPI is great and all but I do not want to add another layer between the controller and the game. And I do not want the controller to depend on steam to function. Of course it still can be passed through SIAPI as a generic xbox controller.
I am planning to develop a companion app for graphical controller configuration in the future, but configs would still be on the controller itself. Also I had a thought about using SIAPI format for configs but it would most likely mean I'd need to develop a parser as well instead of relying on something ready like a json parser. Also I would be bound by SIAPI features instead of developing what I want the most. That support could be added later anyways.
